File: server/utils/siteSettings.js
const path = require('path');
const fs = require('fs/promises');
const SiteSettings = require('../models/SiteSettings');

const DATA_DIR = process.env.ORDER_DATA_DIR
  ? path.resolve(process.env.ORDER_DATA_DIR)
  : path.join(__dirname, '..', 'data');
const SETTINGS_FILE = path.join(DATA_DIR, 'site-settings.json');

/**
 * 預設值 = 既有前端/App 寫死的內容，確保未編輯前行為不變。
 */
const DEFAULT_SETTINGS = {
  shop: {
    name: '磐石烤地瓜 - 番薯阿嬤',
    phone: '0953830409',
    email: 'sweetpotatograndmom@gmail.com',
    address: '高雄市左營區華夏路576號',
  },
  line: {
    officialId: '@437lnypi',
    addFriendUrl: 'https://line.me/R/ti/p/@437lnypi',
  },
  shipping: {
    fee: 150,
    freeThresholdQty: 20,
    freeThresholdKeyword: '黃金地瓜',
  },
  content: {
    heroTag: '傳承古早窯烤工藝',
    heroTitle: '番薯阿嬤的溫暖滋味',
    heroSubtitle: '炭火慢烤，每一口都是古早的人情味。',
    storyTitle: '守護一爐炭火的執著',
    storyBody:
      '在那座用了 20 多年的窯烤爐旁，我們依然堅持最古老的做法。阿嬤常說：「地瓜要好吃，急不得。」每一顆地瓜都承載著我們對土地的敬意。',
  },
};

let writeQueue = Promise.resolve();
function queueWrite(task) {
  writeQueue = writeQueue.then(task, task);
  return writeQueue;
}

function isObject(value) {
  return value && typeof value === 'object' && !Array.isArray(value);
}

/**
 * 深層合併（只覆蓋有提供的欄位），用於套用預設與部分更新。
 */
function deepMerge(base, override) {
  const out = Array.isArray(base) ? [...base] : { ...base };
  if (!isObject(override)) return out;
  for (const [key, value] of Object.entries(override)) {
    if (value === undefined) continue;
    if (isObject(value) && isObject(out[key])) {
      out[key] = deepMerge(out[key], value);
    } else {
      out[key] = value;
    }
  }
  return out;
}

/**
 * 只挑出允許的欄位（防止寫入任意鍵），並做基本型別正規化。
 */
function sanitizePatch(patch) {
  const clean = {};
  if (isObject(patch.shop)) {
    clean.shop = {};
    for (const k of ['name', 'phone', 'email', 'address']) {
      if (patch.shop[k] !== undefined) clean.shop[k] = String(patch.shop[k]).trim();
    }
  }
  if (isObject(patch.line)) {
    clean.line = {};
    for (const k of ['officialId', 'addFriendUrl']) {
      if (patch.line[k] !== undefined) clean.line[k] = String(patch.line[k]).trim();
    }
  }
  if (isObject(patch.shipping)) {
    clean.shipping = {};
    if (patch.shipping.fee !== undefined) {
      clean.shipping.fee = Math.max(0, Number(patch.shipping.fee) || 0);
    }
    if (patch.shipping.freeThresholdQty !== undefined) {
      clean.shipping.freeThresholdQty = Math.max(0, Math.floor(Number(patch.shipping.freeThresholdQty) || 0));
    }
    if (patch.shipping.freeThresholdKeyword !== undefined) {
      clean.shipping.freeThresholdKeyword = String(patch.shipping.freeThresholdKeyword).trim();
    }
  }
  if (isObject(patch.content)) {
    clean.content = {};
    for (const k of ['heroTag', 'heroTitle', 'heroSubtitle', 'storyTitle', 'storyBody']) {
      if (patch.content[k] !== undefined) clean.content[k] = String(patch.content[k]);
    }
  }
  return clean;
}

async function loadFromFile() {
  try {
    await fs.mkdir(DATA_DIR, { recursive: true });
    const raw = await fs.readFile(SETTINGS_FILE, 'utf8');
    const parsed = JSON.parse(raw);
    return isObject(parsed) ? parsed : {};
  } catch {
    return {};
  }
}

async function saveToFile(settings) {
  await fs.mkdir(DATA_DIR, { recursive: true });
  const tmp = `${SETTINGS_FILE}.${process.pid}.${Date.now()}.tmp`;
  await fs.writeFile(tmp, JSON.stringify(settings, null, 2), 'utf8');
  await fs.rename(tmp, SETTINGS_FILE);
}

function stripMeta(doc) {
  if (!doc) return {};
  const plain = typeof doc.toObject === 'function' ? doc.toObject() : doc;
  const { _id, __v, key, createdAt, updatedAt, ...rest } = plain;
  return rest;
}

/**
 * 讀取目前設定（預設值 + 已儲存的覆蓋）。永遠回傳完整結構。
 */
async function getSiteSettings(app) {
  let stored = {};
  if (app?.locals?.db?.ready) {
    try {
      const doc = await SiteSettings.findOne({ key: 'default' }).lean();
      if (doc) stored = stripMeta(doc);
    } catch (err) {
      console.error('讀取 SiteSettings (Mongo) 失敗:', err.message);
    }
  } else {
    stored = await loadFromFile();
  }
  return deepMerge(DEFAULT_SETTINGS, stored);
}

/**
 * 套用部分更新並儲存。回傳更新後的完整設定。
 */
async function updateSiteSettings(app, patch) {
  const clean = sanitizePatch(patch || {});
  const current = await getSiteSettings(app);
  const next = deepMerge(current, clean);

  if (app?.locals?.db?.ready) {
    await SiteSettings.findOneAndUpdate(
      { key: 'default' },
      { $set: { ...next, key: 'default' } },
      { upsert: true, new: true },
    );
  }
  // 一律也寫入檔案（fallback 主要儲存 / Mongo 模式下的備份）
  await queueWrite(() => saveToFile(next));
  return next;
}

module.exports = {
  DEFAULT_SETTINGS,
  getSiteSettings,
  updateSiteSettings,
  SETTINGS_FILE,
};

File: mobile/lib/models/site_settings.dart
/// 店家設定模型（對應後端 /api/site-settings）。
///
/// 預設值與既有寫死內容一致，確保離線 / 連線失敗時行為不變。
class SiteSettings {
  final ShopInfo shop;
  final LineInfo line;
  final ShippingRule shipping;
  final SiteContent content;

  const SiteSettings({
    required this.shop,
    required this.line,
    required this.shipping,
    required this.content,
  });

  /// 預設值（= 之前寫死於 AppConfig / 畫面的內容）。
  static const SiteSettings defaults = SiteSettings(
    shop: ShopInfo(
      name: '磐石烤地瓜 - 番薯阿嬤',
      phone: '0953830409',
      email: 'sweetpotatograndmom@gmail.com',
      address: '高雄市左營區華夏路576號',
    ),
    line: LineInfo(
      officialId: '@437lnypi',
      addFriendUrl: 'https://line.me/R/ti/p/@437lnypi',
    ),
    shipping: ShippingRule(
      fee: 150,
      freeThresholdQty: 20,
      freeThresholdKeyword: '黃金地瓜',
    ),
    content: SiteContent(
      heroTag: '傳承古早窯烤工藝',
      heroTitle: '番薯阿嬤的溫暖滋味',
      heroSubtitle: '炭火慢烤，每一口都是古早的人情味。',
      storyTitle: '守護一爐炭火的執著',
      storyBody:
          '在那座用了 20 多年的窯烤爐旁，我們依然堅持最古老的做法。阿嬤常說：「地瓜要好吃，急不得。」每一顆地瓜都承載著我們對土地的敬意。',
    ),
  );

  factory SiteSettings.fromJson(Map<String, dynamic> json) {
    final d = defaults;
    final shop = json['shop'] as Map<String, dynamic>? ?? const {};
    final line = json['line'] as Map<String, dynamic>? ?? const {};
    final shipping = json['shipping'] as Map<String, dynamic>? ?? const {};
    final content = json['content'] as Map<String, dynamic>? ?? const {};

    String s(dynamic v, String fallback) {
      final str = (v ?? '').toString().trim();
      return str.isEmpty ? fallback : str;
    }

    int i(dynamic v, int fallback) {
      if (v is num) return v.toInt();
      return int.tryParse('${v ?? ''}') ?? fallback;
    }

    return SiteSettings(
      shop: ShopInfo(
        name: s(shop['name'], d.shop.name),
        phone: s(shop['phone'], d.shop.phone),
        email: s(shop['email'], d.shop.email),
        address: s(shop['address'], d.shop.address),
      ),
      line: LineInfo(
        officialId: s(line['officialId'], d.line.officialId),
        addFriendUrl: s(line['addFriendUrl'], d.line.addFriendUrl),
      ),
      shipping: ShippingRule(
        fee: i(shipping['fee'], d.shipping.fee),
        freeThresholdQty: i(shipping['freeThresholdQty'], d.shipping.freeThresholdQty),
        // 關鍵字允許為空字串（代表全部商品都算），因此不套用 fallback。
        freeThresholdKeyword: (shipping.containsKey('freeThresholdKeyword'))
            ? (shipping['freeThresholdKeyword'] ?? '').toString()
            : d.shipping.freeThresholdKeyword,
      ),
      content: SiteContent(
        heroTag: s(content['heroTag'], d.content.heroTag),
        heroTitle: s(content['heroTitle'], d.content.heroTitle),
        heroSubtitle: s(content['heroSubtitle'], d.content.heroSubtitle),
        storyTitle: s(content['storyTitle'], d.content.storyTitle),
        storyBody: s(content['storyBody'], d.content.storyBody),
      ),
    );
  }
}

class ShopInfo {
  final String name;
  final String phone;
  final String email;
  final String address;
  const ShopInfo({
    required this.name,
    required this.phone,
    required this.email,
    required this.address,
  });
}

class LineInfo {
  final String officialId;
  final String addFriendUrl;
  const LineInfo({required this.officialId, required this.addFriendUrl});
}

class ShippingRule {
  final int fee;
  final int freeThresholdQty;
  final String freeThresholdKeyword;
  const ShippingRule({
    required this.fee,
    required this.freeThresholdQty,
    required this.freeThresholdKeyword,
  });
}

class SiteContent {
  final String heroTag;
  final String heroTitle;
  final String heroSubtitle;
  final String storyTitle;
  final String storyBody;
  const SiteContent({
    required this.heroTag,
    required this.heroTitle,
    required this.heroSubtitle,
    required this.storyTitle,
    required this.storyBody,
  });
}
